### PR TITLE
Add swift monad post

### PR DESCRIPTION
**Pull request by endoze**
This commit adds the first draft of the 'Swift Optionals Are Monads
Too!' post.

### CODE REVIEW

**Comment by calebkm** [comment]
w00t w000t! This is great, nice job taking the initiative and pumping this out. I'd say it's still _slightly_ over my head, but I'm not very smart so that's not too surprising.

File: posts/Swift-Optionals-Are-Monads-Too.md [hunk 1]
```diff
@@ -0,0 +1,264 @@
...
+    public func map<U>(@noescape f: (Wrapped) throws -> U) rethrows -> U?
+    /// Returns `nil` if `self` is `nil`, `f(self!)` otherwise.
+    @warn_unused_result
+    public func flatMap<U>(@noescape f: (Wrapped) throws -> U?) rethrows -> U?
+    /// Create an instance initialized with `nil`.
+    public init(nilLiteral: ())
+}
+```
+
+Reading the first comment in the source code gives us an awesome definition of what an Optional really is.
```
Comment (calebkm) [inline posts/Swift-Optionals-Are-Monads-Too.md]: My only comment is that the word `awesome` here is maybe just slightly hyperbolic. Maybe `awesome` is a "cooler" word, but I think the sentence flows a little nicer with a `great` or a `perfect`.

Comment (northband) [inline posts/Swift-Optionals-Are-Monads-Too.md]: Or the word `clear` - "...in the source code gives us a clear definition"

File: posts/Swift-Optionals-Are-Monads-Too.md [hunk 2]
```diff
@@ -0,0 +1,264 @@
...
+    /// Create an instance initialized with `nil`.
+    public init(nilLiteral: ())
+}
+```
+
+Reading the first comment in the source code gives us an awesome definition of what an Optional really is. 
+
+> A type that can represent either a Wrapped value or nil, the absence of a value.
+
+Put in other words, an Optional is a container that wraps the concept of having a value or not.
```
Comment (northband) [inline posts/Swift-Optionals-Are-Monads-Too.md]: `Put` may be optional - it flows well as 'In other words,'.

File: posts/Swift-Optionals-Are-Monads-Too.md [hunk 3]
```diff
@@ -0,0 +1,264 @@
...
+
+So why is this useful? Because Swift is a strongly typed language it makes us code for both cases of an Optional, having a value (.Some) or no value at all (.None). When we declare variables with let or var and make them Optional, we are telling the compiler that our variable may contain a value or not. And any subsequent access to our variable will require us to check that it contains a value before performing actions on the wrapped value.
```
Comment (northband) [inline posts/Swift-Optionals-Are-Monads-Too.md]: This may read better:

So why is this useful? Because Swift is a strongly typed language `that` makes us code for both cases of an Optional.  That is to say, having a value (.Some) or no value at all (.None). When we declare variables with let or var and make them Optional, we are telling the compiler that our variable may contain a value or not. `Any` subsequent access to our variable will require us to check that it contains a value before performing actions on the wrapped value.

File: posts/Swift-Optionals-Are-Monads-Too.md [hunk 4]
```diff
@@ -0,0 +1,264 @@
...
+  print(emptyString)
+} else {
+  // here we can handle the case where the Optional
+  // doesn't contain a value
+  print("Your string was empty")
+}
+```
+---
+
+# What the Functor
```
Comment (northband) [inline posts/Swift-Optionals-Are-Monads-Too.md]: May want to add a question mark: What the Functor?

File: posts/Swift-Optionals-Are-Monads-Too.md (hunk 1 above)
Comment (endoze) [inline posts/Swift-Optionals-Are-Monads-Too.md]: I'm just that excited about the definition. I could run with `great`, `perfect`, or `clear`.

File: posts/Swift-Optionals-Are-Monads-Too.md [hunk 5]
```diff
@@ -0,0 +1,264 @@
...
+
+So what is a functor? It sounds complicated and only adding to it's complicated nature, it's roots are in Mathematics (Category Theory to be a bit more specific). But let's try and define it in terms a programmer understands instead of mathematical terms.
```
Comment (northband) [inline posts/Swift-Optionals-Are-Monads-Too.md]: Maybe 'adding to it's `complex` nature,...'

File: posts/Swift-Optionals-Are-Monads-Too.md [hunk 6]
```diff
@@ -0,0 +1,264 @@
...
+public enum Optional<T> {
+    // ...
+
+    public func map<U>(f: (T) -> U) -> U?
+
+    // ...
+}
+```
+
+If we break down Optional's implementation of map, it's a function that accepts another function as it's only argument. This other function, we'll call it f, accepts an object of the type T and returns an object of the type U. And finally, map returns U? which is an Optional of the type U.
```
Comment (northband) [inline posts/Swift-Optionals-Are-Monads-Too.md:77]: 👍

File: posts/Swift-Optionals-Are-Monads-Too.md [hunk 7]
```diff
@@ -0,0 +1,264 @@
...
+
+--- 
+
+# Applicative Functor
+
+Armed with the knowledge of what a Functor is, next up is Applicative Functor. Let's start with a definition in programmer terms.
+
+> An applicative functor is any type that defines apply.
+
+Looking back at the declaration for Optional, we don't find any methods defining apply. Well hmm, by default Optionals aren't applicative functors. But wait! We can add an extension to this built in type and make it work.
```
Comment (northband) [inline posts/Swift-Optionals-Are-Monads-Too.md]: It's totally acceptable but you don't need to use words like 'Well hmm,' you can just do 'Hmm'.  Or 'But wait!', could be 'Wait!'.  Again this is just a style thing.

File: posts/Swift-Optionals-Are-Monads-Too.md [hunk 8]
```diff
@@ -0,0 +1,264 @@
...
+
+```swift
+extension Optional {
+  func apply<U>(f: (Wrapped -> U)?) -> U? {
+    return f.flatMap { self.map($0) }
+  }
+}
+```
+
+Looking at the implementation of apply kind of melts your brain, so let's break it down and explain it in simpler terms. First we should notice a very important detail buried in the middle of the function. f's type is (Wrapped -> U)?. f is an Optional! This little interesting tidbit is the key difference between map and apply. So looking into the implementation we see that we are calling flatMap on f. So if f is not nil we are going to call self.map($0) where $0 is referring to f. So putting it all together, we pass in an Optional function to apply. If the function exists, we call map on our Optional and pass in our function. So if either our function is nil or our Optionals value is nil, we return nil. If both are not nil, we return the result of calling our function with the value wrapped in the Optional.
```
Comment (northband) [inline posts/Swift-Optionals-Are-Monads-Too.md:126]: 👍

File: posts/Swift-Optionals-Are-Monads-Too.md [hunk 9]
```diff
@@ -0,0 +1,264 @@
...
+// have been set or initialized
+
+var friendsStatus = numberOfFriends.map { 
+  "You have \($0) friends."
+} ?? "No friends yet." 
+```
+
+This allows us to choose two completely different strings based on whether our `numberOfFriends` variable has a value or not.
+
+One last example of a great usage for `map`. What if we want to read the contents of a url?
```
Comment (northband) [inline posts/Swift-Optionals-Are-Monads-Too.md:200]: Maybe:

Another example of a great usage...

This may cut down on simple minds thinking this is not the actual last case where it's useful ;-)

File: posts/Swift-Optionals-Are-Monads-Too.md [hunk 10]
```diff
@@ -0,0 +1,264 @@
...
+var webContents = webURL.map {
+  NSData(contentsOfURL: $0)
+}.map {
+  NSString(data: $0, encoding: NSUTF8StringEncoding)
+}
+
+webContents.map { print($0) }
+```
+
+If I was a bit more mischievous, I could leave the example like that and go on with the post. But the most astute reader who really gets these concepts would point out the mistake in the code sample above. In case you didn't catch it, that's ok. The problem with our code above is that map accepts a closure/function that does not return an `Optional`. `NSData`'s initializer that accepts an `NSURL` is a failable initializer. Since it can return `nil`, our closure can return `nil`. Xcode will complain inside our second call to `map` that `$0` has not been unwrapped. So what do we do?
```
Comment (northband) [inline posts/Swift-Optionals-Are-Monads-Too.md:214]: 👍

File: posts/Swift-Optionals-Are-Monads-Too.md [hunk 11]
```diff
@@ -0,0 +1,264 @@
...
+var webContents = webURL.flatMap {
+  NSData(contentsOfURL: $0)
+}.flatMap {
+  NSString(data: $0, encoding: NSUTF8StringEncoding)
+}
+
+webContents.map { print($0) }
+```
+
+### How do we really apply
```
Comment (northband) [inline posts/Swift-Optionals-Are-Monads-Too.md:229]: Maybe:

Real world uses of apply

File: posts/Swift-Optionals-Are-Monads-Too.md [hunk 12]
```diff
@@ -0,0 +1,264 @@
...
+
+  return url.flatMap {
+    NSData(contentsOfURL: $0)
+  }.apply(parseResponseHandler)
+}
+```
+
+# Wrapping Up
+
+Hopefully by now you can see how useful functors, applicative functors, and monads can really be. Apple making Optionals work as functors and monads really make writing intentful code a cinch. And just to keep your brain thinking about this crazy topic, Swift's Array type is also a functor and a monad!
```
Comment (northband) [inline posts/Swift-Optionals-Are-Monads-Too.md]: Maybe:

Apple's decision to make Optionals work as functors and monads really allows writing confident code a cinch. Lastly, just to keep your brain thinking about this crazy topic, Swift's Array type is also a functor and a monad!

**Comment by northband** [comment]
This is really good @endoze - nice work!  I just made red marks to shorten things up a bit.  Take them as you will.

File: posts/Swift-Optionals-Are-Monads-Too.md (hunk 12 above)
Comment (endoze) [inline posts/Swift-Optionals-Are-Monads-Too.md]: I'm no fan of using `lastly`, otherwise I like the suggested edits.

File: posts/Swift-Optionals-Are-Monads-Too.md (hunk 11 above)
Comment (endoze) [inline posts/Swift-Optionals-Are-Monads-Too.md:229]: I was aiming for a little bit on the light side instead of sticking with something specific and to the point.
